### PR TITLE
[core] Validate changelog producer on table without primary keys

### LINKED ISSUE #1782: [Feature] Validate write-mode: auto and changelog producer not NONE conflict if no primary key
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

Now we have table schema validation on WriteMode.APPEND_ONLY  and changelog-produer not NONE. This can throw runtime exception when users set the writeMode to APPEND_ONLY explicitly. 

But when there is a table without primary key and user set the changelog-producer to something else but not NONE, this validation is passed at first because the writeMode is AUTO  when validating. Then paimon will help users to set the writeMode to APPEND_ONLY  because there is no primary key in table.

Thus we have a table without primary key, with writeMode APPEND_ONLY and changelog-producer != NONE. This will cause the reader waiting for changelog files but the writer doesn't have any changelog to write. 

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by schnappi17**
### Purpose

Linked issue: close #1782

### CODE REVIEW

**Comment by schnappi17** [comment]
@JingsongLi Could you help to review this pr?

File: paimon-core/src/main/java/org/apache/paimon/schema/SchemaValidation.java [hunk 1]
```diff
@@ -92,6 +92,15 @@ public static void validateTableSchema(TableSchema schema) {
                             WRITE_MODE.key(), APPEND_ONLY, CHANGELOG_PRODUCER.key()));
         }
 
+        if (options.writeMode() == WriteMode.AUTO
+                && schema.primaryKeys().isEmpty()
```
Comment (SteNicholas) [inline paimon-core/src/main/java/org/apache/paimon/schema/SchemaValidation.java:96]:
```suggestion
                && CollectionUtil.isNullOrEmpty(schema.primaryKeys())
```

File: paimon-core/src/main/java/org/apache/paimon/schema/SchemaValidation.java [hunk 2]
```diff
@@ -92,6 +92,15 @@ public static void validateTableSchema(TableSchema schema) {
                             WRITE_MODE.key(), APPEND_ONLY, CHANGELOG_PRODUCER.key()));
         }
 
+        if (options.writeMode() == WriteMode.AUTO
```
Comment (SteNicholas) [inline paimon-core/src/main/java/org/apache/paimon/schema/SchemaValidation.java:95]: Should we check the write mode?

File: paimon-core/src/main/java/org/apache/paimon/schema/SchemaValidation.java [hunk 3]
```diff
@@ -92,6 +92,15 @@ public static void validateTableSchema(TableSchema schema) {
...
+                && CollectionUtil.isNullOrEmpty(schema.primaryKeys())
+                && changelogProducer != ChangelogProducer.NONE) {
+            throw new UnsupportedOperationException(
+                    String.format(
+                            "Can not set %s on table without primary keys.",
```
Comment (SteNicholas) [inline paimon-core/src/main/java/org/apache/paimon/schema/SchemaValidation.java:100]: We could improve the exception message to help users to solve, like `when set write mode to none, the table must have primary key` etc. Meanwhile, adds this information into document to reduce the situation that the users have this validation failure.

**Review by SteNicholas — APPROVED** [review]
LGTM.

**Comment by schnappi17** [comment]
Error:  Failures: 
Error:    MySqlSyncDatabaseTableListITCase.testActionRunResult:133

**Comment by yuzelin** [comment]
I am trying to fix that error, I'll ping you after I fix it.

**Pull request merged**